smart_card_connector_app/src/chrome-api-provider.js

```diff
@@ -0,0 +1,165 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+goog.provide('GoogleSmartCard.ConnectorApp.ChromeApiProvider');
+
+goog.require('GoogleSmartCard.PcscLiteClient.API');
+goog.require('GoogleSmartCard.PcscLiteServerClientsManagement.ServerRequestHandler');
+goog.require('goog.Disposable');
+goog.require('goog.log');
+goog.require('goog.log.Logger');
+goog.require('goog.messaging.AbstractChannel');
+
+goog.scope(function() {
+
+const GSC = GoogleSmartCard;
+const Pcsc = GSC.PcscLiteServerClientsManagement;
+const PcscApi = GSC.PcscLiteClient.API;
+
+/** @type {!goog.log.Logger} */
+const logger = GSC.Logging.getScopedLogger('ConnectorApp.ChromeApiProvider');
+
+/**
+ * Returns the result code for chrome.smartCardProviderPrivate API that
+ * corresponds to the given PCSC error.
+ * @param {!PcscApi.ERROR_CODE} errorCode
+ * @return {!chrome.smartCardProviderPrivate.ResultCode}
+ * @throws {Error} Will throw if unknown error code encountered.
+ */
+function convertErrorCodeToEnum(errorCode) {
+  switch (errorCode) {
+    case PcscApi.SCARD_F_INTERNAL_ERROR:
+      return chrome.smartCardProviderPrivate.ResultCode.INTERNAL_ERROR;
+    // TODO(vkovalova): add other error codes
+    default:
+      throw new Error(`Unknown error code ${errorCode} encountered`);
+  }
+}
+
+/**
+ * This class subscribes to a Chrome event with a callback, specified in the
+ * constructor arguments. Unsubscribes on disposal.
+ */
+class ChromeEventListener extends goog.Disposable {
+  constructor(chromeEvent, callback) {
+    super();
+    this.chromeEvent_ = chromeEvent;
+    this.callback_ = callback;
+    this.chromeEvent_.addListener(this.callback_);
+  }
+
+  /** @override */
+  disposeInternal() {
+    this.chromeEvent_.removeListener(this.callback_);
+  }
+};
+
+/**
+ * This class provides chrome.smartCardProviderPrivate API with PCSC responses.
+ *
+ * On creation, it subscribes to chrome.smartCardProviderPrivate API events,
+ * which correspond to different PCSC requests. When an event is fired,
+ * ChromeApiProvider sends a request to Pcsc.ServerRequestHandler and waits for
+ * the response. When the response is received, it reports the result back using
+ * chrome.smartCardProviderPrivate API.
+ */
+GSC.ConnectorApp.ChromeApiProvider = class extends goog.Disposable {
+  /**
+   * @param {!goog.messaging.AbstractChannel} serverMessageChannel
+   * @param {!Pcsc.ReadinessTracker} serverReadinessTracker
+   */
+  constructor(serverMessageChannel, serverReadinessTracker) {
+    super();
+    this.serverRequester_ = new Pcsc.ServerRequestHandler(
+        serverMessageChannel, serverReadinessTracker, 'chrome');
+
+    this.chromeEventListeners_ = [];
```

Comment:
nit: Add JSDoc that documents it's an array of `ChromeEventListener`.

Reply:
done.